Updated Kernel GPIO driver doc with examples and command-line tools

jonaswood01 opened the pull request:
In response to the Linux Kernel GPIO in the SDK Docs feeling insufficient. So, this update walks someone through a quick example for using GPIO in user space and provides resources for command-line tools and additional guides.

praneethbajjuri commented:
@DhruvaG2000 for review

praneethbajjuri commented:
@jonaswood01  thanks for sending this PR. Review the checker failures. https://github.com/TexasInstruments/processor-sdk-doc/pull/234/checks

madweshanshu commented on source/linux/Foundational_Components/Kernel/Kernel_Drivers/GPIO.rst:
```diff
@@ -48,33 +46,186 @@ and interrupt generation.
+
+      - Linux:  https://e2e.ti.com/support/processors-group/processors/f/processors-forum/1494485/faq-processor-sdk-getting-started-with-gpios-in-linux-using-the-am62-and-am64-family-processors
+      - U-Boot: https://e2e.ti.com/support/processors-group/processors/f/processors-forum/1398803/faq-processor-sdk-am62x-how-to-toggle-gpios-and-leds-from-u-boot-command-prompt
+      - GPIO on Sitara: https://www.ti.com/lit/ab/spradk0/spradk0.pdf?ts=1743432396299&ref_url=https%253A%252F%252Fwww.google.com%252F
```
Remove the question mark and all the characters after the question mark. 
Its makes the url cleaner and the delivery information isn't important.

madweshanshu commented on source/linux/Foundational_Components/Kernel/Kernel_Drivers/GPIO.rst:
```diff
@@ -48,33 +46,186 @@ and interrupt generation.
+
+- This should result in a large output, 168+ lines in this case. It outputs all GPIO lines available on every GPIO chip, as well as their names and input/output directions. The above output is truncated.
+
+- For more info on *gpioinfo*, see its man page: https://manpages.debian.org/experimental/gpiod/gpioinfo.1.en.html.
```
Is there a reason to link the Debian page rather than the 'libgpiod.readthedocs.io' command page? If not, then lets link the readthedocs page.

jonaswood01 replied:
Great point, will remove Debian links and replace with lgpiod links

madweshanshu reviewed (changes requested):
Looks good! I think the man page links should refer to the source rather than the Debian page.

Thank you for linking the GPIO E2E FAQs :)

cshilwant commented on source/linux/Foundational_Components/Kernel/Kernel_Drivers/GPIO.rst:
```diff
@@ -48,33 +47,186 @@ and interrupt generation.
+
+- For more info on *gpioset*, see its man page: https://manpages.debian.org/experimental/gpiod/gpioset.1.en.html.
+
+For more command-line tools, refer to the *libgpiod* documentation: https://libgpiod.readthedocs.io/en/latest/gpio_tools.html and the *gpiod* Debian documentation: https://manpages.debian.org/experimental/gpiod/index.html.
```
Any reason to point to gpiod package from debian.org in yocto docs?

jonaswood01 replied:
Will change links to lgpiod only.

StaticRocket commented on source/linux/Foundational_Components/Kernel/Kernel_Drivers/GPIO.rst:
```diff
@@ -48,33 +47,186 @@ and interrupt generation.
+The GPIO driver can be used in user space via Linux CLI. This section provides examples, command-line-tools, and guidance for using GPIOs in user space.
 
 .. note ::
-   Since linux 4.8 the GPIO sysfs interface is deprecated. User space should use
```
Style nitpick: newline after directive

StaticRocket commented on source/linux/Foundational_Components/Kernel/Kernel_Drivers/GPIO.rst:
```diff
@@ -48,33 +47,186 @@ and interrupt generation.
+   Since linux 4.8, the GPIO sysfs interface is deprecated. User space should use
    the character device instead.
 
+# Example: Using GPIOs with the *libgpiod* library in user space to toggle a GPIO pin, which is connected to an LED on the SK-AM62B-P1 board:
```
```suggestion
Example: Using GPIOs with the *libgpiod* library in user space to toggle a GPIO pin, which is connected to an LED on the SK-AM62B-P1 board:
```

StaticRocket commented on source/linux/Foundational_Components/Kernel/Kernel_Drivers/GPIO.rst:
```diff
@@ -48,33 +47,186 @@ and interrupt generation.
+
+#. Locate *Table 2-23* under **2.6.16.3 User Test LEDs** in the *AM62x Starter Kit User's Guide (Rev. C)*: https://www.ti.com/lit/pdf/spruj40.
+
+- This example uses LED "LD11", AKA the **IO_EXP_TEST_LED**. It is connected to GPIO *U70.24(P27)*, AKA pin 27 of the *GPIO Port Expander* on the SK-AM62B-P1 board:
```
Mixed list and numbers? Why? If you are trying to nest sections use the proper indentation.

StaticRocket commented on source/linux/Foundational_Components/Kernel/Kernel_Drivers/GPIO.rst:
```diff
@@ -48,33 +47,186 @@ and interrupt generation.
+
+- This example uses LED "LD11", AKA the **IO_EXP_TEST_LED**. It is connected to GPIO *U70.24(P27)*, AKA pin 27 of the *GPIO Port Expander* on the SK-AM62B-P1 board:
+
+.. Image:: /images/sk-am62b-p1-top.png
```
Should at least define a width or height here so the image doesn't take up a large section of the screen.

StaticRocket commented on source/linux/Foundational_Components/Kernel/Kernel_Drivers/GPIO.rst:
```diff
@@ -48,33 +47,186 @@ and interrupt generation.
+
+   .. code-block:: console
+
+         $ gpiodetect
```
Incorrect indentation depth

StaticRocket commented on source/linux/Foundational_Components/Kernel/Kernel_Drivers/GPIO.rst:
```diff
@@ -48,33 +47,186 @@ and interrupt generation.
+
+#. Read info for every available **gpiochip**:
+
+      .. code-block:: console
```
Incorrect indentation depth

StaticRocket replied:
This is the case for a number of other code blocks in this file as well.

StaticRocket commented on source/linux/Foundational_Components/Kernel/Kernel_Drivers/GPIO.rst:
```diff
@@ -48,33 +47,186 @@ and interrupt generation.
+         $ gpioset -c 2 23=0
+         ^C
+
+- *Note: '^C' (Ctrl + C) interrupts the command, else the command would not terminate on its own. Use '^C' (Ctrl + C) to stop the command, and the new GPIO value will still remain.*
```
This note is incorrect.

https://manpages.debian.org/experimental/gpiod/gpioset.1.en.html#NOTE

makohoek reviewed (changes requested):
Hi @jonaswood01,

Thank you for the patch. Looking at the authorship, it seems you are not the author of this patch:
```
commit 1eb41f7a21f9bb2034c74252f885a71e2ec27cec (jonaswood01/master, jonaswood01/HEAD)
Author: Maxime Fournier <mfournier@baylibre.com>
Date:   Tue Apr 1 16:52:17 2025 +0200

    docs(GPIO): Updated Kernel driver doc with examples
    
    Replaced Debian links with lgpiod links and cleaned a lengthy URL.
    
    Signed-off-by: Jonas Wood <jonaslwood@gmail.com>
```

Can you please fix this? It seems weird that Maxime is the author here since you are the one submitting the pull request.
This can be fixed with:
```
$ git commit --amend --reset-author
```

Also, can you rebase this, please? It seems that this commit modifies some Android files (such as `source/android/Foundational_Components_Bootloaders.rst`) which seem unrelated to this change.

See: https://docs.github.com/en/get-started/using-git/about-git-rebase for help with `git rebase`

jonaswood01 commented:
> > Hi @jonaswood01,
> > Thank you for the patch. Looking at the authorship, it seems you are not the author of this patch:
> > ```
> > commit 1eb41f7a21f9bb2034c74252f885a71e2ec27cec (jonaswood01/master, jonaswood01/HEAD)
> > Author: Maxime Fournier <mfournier@baylibre.com>
> > Date:   Tue Apr 1 16:52:17 2025 +0200
> > 
> >     docs(GPIO): Updated Kernel driver doc with examples
> >     
> >     Replaced Debian links with lgpiod links and cleaned a lengthy URL.
> >     
> >     Signed-off-by: Jonas Wood <jonaslwood@gmail.com>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Can you please fix this? It seems weird that Maxime is the author here since you are the one submitting the pull request. This can be fixed with:
> > ```
> > $ git commit --amend --reset-author
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Also, can you rebase this, please? It seems that this commit modifies some Android files (such as `source/android/Foundational_Components_Bootloaders.rst`) which seem unrelated to this change.
> > See: https://docs.github.com/en/get-started/using-git/about-git-rebase for help with `git rebase`

It should be fixed now

jmenti commented on source/linux/Foundational_Components/Kernel/Kernel_Drivers/GPIO.rst:
```diff
@@ -162,7 +162,11 @@ Example: Using GPIOs with the *libgpiod* library in user space to toggle a GPIO
       $ gpioset -c 2 23=0
       ^C
 
-   - *Note: '^C' (Ctrl + C) interrupts the command, else the command would not terminate on its own. Use '^C' (Ctrl + C) to stop the command, and the new GPIO value will still remain.*
```
Why not include this second patch with your first patch?

jonaswood01 replied:
Good catch, that second patch edit was meant to be a part of the first patch. It's now consolidated into one patch.

makohoek reviewed (approved):
I mostly review Android related things (which are no longer present in this commit).

I confirm that:
- commit authorship looks good
- there are no longer any (unrelated) Android changes in this PR.

I have not thouroughly reviewed anything else, but approving anyways.
